### DISCUSSION TITLE
✅ Mock `options` in test

### DISCUSSION THREAD

**Discussion by orimdominic** [RobinTail/express-zod-api#1039] [Ideas]
Hello @RobinTail,

I must say that I am impressed by the developer experience that this library provides. Oh! I wish that this was Express itself.
I have also gone through pull requests and issues and oh my! Your work rate on this is amazing. I appreciate that!

I would like to know if we can create a mock for the `options` object that is passed to a request handler in tests, the same way we can for `request` and `logger` using `requestProps` and `loggerProps` respectively.

A use case will be for a handler that attaches a method to `options` that sends emails,  but during tests, we want this method to be mocked to avoid making requests to the mailing server. Let me show you what I mean in code:

We have this endpoint:
```typescript
export const sendEmailEndpoint = defaultEndpointsFactory
  .addOptions({
    sendMail: (email: string, content: string) =>
      console.info(`sending ${content} to ${email}`),
  })
  .build({
    method: "post",
    input: z.object({
      email: z.string(),
    }),
    output: z.object({
      message: z.string(),
    }),
    handler: async ({ input, options, logger }) => {
      options.sendMail(input.email, "Lorem ipsum");
      return { message: `Email sent successfully` };
    },
  });
```

We test it

```ts
describe("sendEmailEndpoint", () => {
  it("should respond successfully", async () => {
    const testEmail = "test@email.com"
    const { responseMock, optionsMock } = await testEndpoint({
      endpoint: sendEmailEndpoint,,
      optionsProps: {
        sendEmail: jest.fn()
      },
      requestProps: {
        method: "POST",
        body: {
          email: testEmail
        },
      },
    });
    expect(optionsMock.sendMail).toBeCalledWith(testEmail, "Lorem ipsum");
    expect(responseMock.status).toBeCalledWith(200);
    expect(responseMock.json).toBeCalledWith({
      status: "success",
      data: { message: `Email sent successfully` },
    });
  });
});
```

We can also mock `options.sendMail` to reject or throw in separate test and then test that an error message and an error status code were returned.

Is there any way we can make this work? I am willing to collaborate to see it come to life.

**Comment by RobinTail** [comment]
Hello @orimdominic ,

thank you for appreciating the efforts of me and contributors.

In order to achieve your desire of mocking the email sending function,
You need to keep that function in a separate module (another file).

In your test, before importing the endpoint, you should mock the module containing that function. 

Please refer to Jest documentation on own modules mocking:

https://jestjs.io/docs/manual-mocks#mocking-user-modules

And StackOverflow. 

Use the `jest.mock()` method. 

Without `__mocks__` dir it should be similar to following:

```ts
const sendMail = jest.fn();
jest.mock('./mailUtil', () => ({ sendMail }));

import { myEndpoint } from './endpoint';

// in test:

expect(sendMail).toHaveBeenCalledWith();
```

Something like that.
I hope this information will help you.

**Reply by orimdominic** [reply]
Yes it does.
Thank you.

**Comment by RobinTail** [comment]
🚀 The whole thing about `testEndpoint` was simplified in the recently released v20, @orimdominic 
No mocking required anymore.

**Reply by orimdominic** [reply]
Thanks for this. I'll look into it 💪

**Comment by RobinTail** [comment]
@orimdominic , there is now `testMiddleware` available for convenience
